File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.java
package org.wikipedia.dataclient;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.collection.LruCache;

import org.apache.commons.lang3.StringUtils;
import org.wikipedia.WikipediaApp;
import org.wikipedia.analytics.eventplatform.DestinationEventService;
import org.wikipedia.analytics.eventplatform.EventService;
import org.wikipedia.analytics.eventplatform.StreamConfig;
import org.wikipedia.dataclient.okhttp.OkHttpConnectionFactory;
import org.wikipedia.json.GsonUtil;
import org.wikipedia.settings.Prefs;

import java.io.IOException;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static org.wikipedia.settings.Prefs.getEventPlatformIntakeUriOverride;

public final class ServiceFactory {
    private static final int SERVICE_CACHE_SIZE = 8;
    private static LruCache<Long, Service> SERVICE_CACHE = new LruCache<>(SERVICE_CACHE_SIZE);
    private static LruCache<Long, RestService> REST_SERVICE_CACHE = new LruCache<>(SERVICE_CACHE_SIZE);
    private static LruCache<Long, CoreRestService> CORE_REST_SERVICE_CACHE = new LruCache<>(SERVICE_CACHE_SIZE);
    private static LruCache<String, EventService> ANALYTICS_REST_SERVICE_CACHE = new LruCache<>(SERVICE_CACHE_SIZE);

    public static Service get(@NonNull WikiSite wiki) {
        long hashCode = wiki.hashCode();
        if (SERVICE_CACHE.get(hashCode) != null) {
            return SERVICE_CACHE.get(hashCode);
        }
        Retrofit r = createRetrofit(wiki, getBasePath(wiki));
        Service s = r.create(Service.class);
        SERVICE_CACHE.put(hashCode, s);
        return s;
    }

    public static RestService getRest(@NonNull WikiSite wiki) {
        long hashCode = wiki.hashCode();
        if (REST_SERVICE_CACHE.get(hashCode) != null) {
            return REST_SERVICE_CACHE.get(hashCode);
        }
        Retrofit r = createRetrofit(wiki, getRestBasePath(wiki));
        RestService s = r.create(RestService.class);
        REST_SERVICE_CACHE.put(hashCode, s);
        return s;
    }

    public static CoreRestService getCoreRest(@NonNull WikiSite wiki) {
        long hashCode = wiki.hashCode();
        if (CORE_REST_SERVICE_CACHE.get(hashCode) != null) {
            return CORE_REST_SERVICE_CACHE.get(hashCode);
        }
        Retrofit r = createRetrofit(wiki, wiki.url() + "/" + CoreRestService.CORE_REST_API_PREFIX);
        CoreRestService s = r.create(CoreRestService.class);
        CORE_REST_SERVICE_CACHE.put(hashCode, s);
        return s;
    }

    public static EventService getAnalyticsRest(@NonNull StreamConfig streamConfig) {
        DestinationEventService destinationEventService = streamConfig.getDestinationEventService();
        if (destinationEventService == null) {
            destinationEventService = DestinationEventService.ANALYTICS;
        }
        if (ANALYTICS_REST_SERVICE_CACHE.get(destinationEventService.getId()) != null) {
            return ANALYTICS_REST_SERVICE_CACHE.get(destinationEventService.getId());
        }

        String intakeBaseUriOverride = getEventPlatformIntakeUriOverride();

        Retrofit r = createRetrofit(null, StringUtils.defaultString(intakeBaseUriOverride, destinationEventService.getBaseUri()));
        EventService s = r.create(EventService.class);
        ANALYTICS_REST_SERVICE_CACHE.put(destinationEventService.getId(), s);
        return s;
    }

    public static <T> T get(@NonNull WikiSite wiki, @Nullable String baseUrl, Class<T> service) {
        Retrofit r = createRetrofit(wiki, TextUtils.isEmpty(baseUrl) ? wiki.url() + "/" : baseUrl);
        return r.create(service);
    }

    private static String getBasePath(@NonNull WikiSite wiki) {
        return TextUtils.isEmpty(Prefs.getMediaWikiBaseUrl()) ? wiki.url() + "/" : Prefs.getMediaWikiBaseUrl();
    }

    public static String getRestBasePath(@NonNull WikiSite wiki) {
        String path = TextUtils.isEmpty(Prefs.getRestbaseUriFormat())
                ? wiki.url() + "/" + RestService.REST_API_PREFIX
                : String.format(Prefs.getRestbaseUriFormat(), "https", wiki.authority());
        if (!path.endsWith("/")) {
            path += "/";
        }
        return path;
    }

    private static Retrofit createRetrofit(@Nullable WikiSite wiki, @NonNull String baseUrl) {
        OkHttpClient.Builder okHttpClientBuilder = OkHttpConnectionFactory.getClient().newBuilder();
        if (wiki != null) {
            okHttpClientBuilder.addInterceptor(new LanguageVariantHeaderInterceptor(wiki));
        }
        return new Retrofit.Builder()
                .client(okHttpClientBuilder.build())
                .baseUrl(baseUrl)
                .client(OkHttpConnectionFactory.getClient().newBuilder().addInterceptor(new LanguageVariantHeaderInterceptor(wiki)).build())
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create(GsonUtil.getDefaultGson()))
                .build();
    }

    private ServiceFactory() { }

    private static class LanguageVariantHeaderInterceptor implements Interceptor {
        @NonNull private final WikiSite wiki;

        LanguageVariantHeaderInterceptor(@NonNull WikiSite wiki) {
            this.wiki = wiki;
        }

        @Override
        public Response intercept(@NonNull Interceptor.Chain chain) throws IOException {
            Request request = chain.request();

            // TODO: remove when the https://phabricator.wikimedia.org/T271145 is resolved.
            if (!request.url().encodedPath().contains("/page/related")) {
                request = request.newBuilder()
                        .header("Accept-Language", WikipediaApp.getInstance().getAcceptLanguage(wiki))
                        .build();
            }

            return chain.proceed(request);
        }
    }
}

File: app/src/main/java/org/wikipedia/dataclient/ServiceFactory.kt
package org.wikipedia.dataclient

import androidx.collection.lruCache
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import org.wikipedia.WikipediaApp
import org.wikipedia.analytics.eventplatform.DestinationEventService
import org.wikipedia.analytics.eventplatform.EventService
import org.wikipedia.analytics.eventplatform.StreamConfig
import org.wikipedia.dataclient.okhttp.OkHttpConnectionFactory
import org.wikipedia.donate.GooglePayComponent
import org.wikipedia.json.JsonUtil
import org.wikipedia.settings.Prefs
import retrofit2.Retrofit
import retrofit2.converter.kotlinx.serialization.asConverterFactory
import retrofit2.create
import java.io.IOException
import java.util.concurrent.TimeUnit

object ServiceFactory {

    private const val SERVICE_CACHE_SIZE = 8

    private val SERVICE_CACHE = lruCache<WikiSite, Service>(SERVICE_CACHE_SIZE, create = {
        // This method is called in the get() method if a value does not already exist.
        createRetrofit(it, getBasePath(it)).create<Service>()
    })

    private val REST_SERVICE_CACHE = lruCache<WikiSite, RestService>(SERVICE_CACHE_SIZE, create = {
        createRetrofit(it, getRestBasePath(it)).create<RestService>()
    })

    private val CORE_REST_SERVICE_CACHE = lruCache<WikiSite, CoreRestService>(SERVICE_CACHE_SIZE, create = {
        createRetrofit(it, it.url() + "/" + CoreRestService.CORE_REST_API_PREFIX).create<CoreRestService>()
    })

    private val ANALYTICS_REST_SERVICE_CACHE = lruCache<DestinationEventService, EventService>(SERVICE_CACHE_SIZE, create = {
        val intakeBaseUriOverride = Prefs.eventPlatformIntakeUriOverride.orEmpty().ifEmpty { it.baseUri }
        createRetrofit(null, intakeBaseUriOverride).create<EventService>()
    })

    private var DONATE_SERVICE_CACHE: Service? = null

    fun get(wiki: WikiSite): Service {
        return SERVICE_CACHE[wiki]!!
    }

    fun getRest(wiki: WikiSite): RestService {
        return REST_SERVICE_CACHE[wiki]!!
    }

    fun getCoreRest(wiki: WikiSite): CoreRestService {
        return CORE_REST_SERVICE_CACHE[wiki]!!
    }

    fun getAnalyticsRest(streamConfig: StreamConfig): EventService {
        return ANALYTICS_REST_SERVICE_CACHE[streamConfig.destinationEventService]!!
    }

    fun getDonate(): Service {
        if (DONATE_SERVICE_CACHE == null) {
            val wikiSite = WikiSite(GooglePayComponent.PAYMENTS_API_URL)
            // https://phabricator.wikimedia.org/T412059
            // Explicitly increase the timeout for the donation API, since the payment processor
            // could occasionally take an increased amount of time to process transactions.
            DONATE_SERVICE_CACHE = createRetrofit(wikiSite, getBasePath(wikiSite), readTimeoutSec = 60L).create<Service>()
        }
        return DONATE_SERVICE_CACHE!!
    }

    operator fun <T> get(wiki: WikiSite, baseUrl: String?, service: Class<T>): T {
        val r = createRetrofit(wiki, baseUrl.orEmpty().ifEmpty { wiki.url() + "/" })
        return r.create(service)
    }

    private fun getBasePath(wiki: WikiSite): String {
        var path = wiki.url()
        if (!path.endsWith("/")) {
            path += "/"
        }
        return path
    }

    fun getRestBasePath(wiki: WikiSite): String {
        var path = if (Prefs.restbaseUriFormat.isEmpty()) wiki.url() + "/" + RestService.REST_API_PREFIX
        else String.format(Prefs.restbaseUriFormat, "https", wiki.authority())
        if (!path.endsWith("/")) {
            path += "/"
        }
        return path
    }

    private fun createRetrofit(wiki: WikiSite?, baseUrl: String, readTimeoutSec: Long = OkHttpConnectionFactory.DEFAULT_READ_TIMEOUT_SEC): Retrofit {
        val builder = OkHttpConnectionFactory.client.newBuilder()
        builder.readTimeout(readTimeoutSec, TimeUnit.SECONDS)
        builder.interceptors().add(builder.interceptors().indexOfFirst { it is HttpLoggingInterceptor }, LanguageVariantHeaderInterceptor(wiki))

        return Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(builder.build())
            .addConverterFactory(JsonUtil.json.asConverterFactory("application/json".toMediaType()))
            .build()
    }

    private class LanguageVariantHeaderInterceptor(private val wiki: WikiSite?) : Interceptor {
        @Throws(IOException::class)
        override fun intercept(chain: Interceptor.Chain): Response {
            var request = chain.request().newBuilder()
                .header("Accept-Language", WikipediaApp.instance.getAcceptLanguage(wiki))
                .build()
            return chain.proceed(request)
        }
    }
}
